src/gpio/convert.rs

```diff
@@ -0,0 +1,372 @@
+        _pupdr: &mut PUPDR<P>,
+    ) -> Pin<Input<PullDown>, HL, P, N> {
+        self.mode::<Input<PullDown>>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as a pulled up input pin
+    pub fn into_pull_up_input(
+        mut self,
+        _moder: &mut MODER<P>,
+        _pupdr: &mut PUPDR<P>,
+    ) -> Pin<Input<PullUp>, HL, P, N> {
+        self.mode::<Input<PullUp>>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as an open drain output pin
+    /// Initial state will be low.
+    pub fn into_open_drain_output(
+        mut self,
+        _moder: &mut MODER<P>,
+        _otyper: &mut OTYPER<P>,
+    ) -> Pin<Output<OpenDrain>, HL, P, N> {
+        self.mode::<Output<OpenDrain>>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as an open-drain output pin.
+    /// `initial_state` specifies whether the pin should be initially high or low.
+    pub fn into_open_drain_output_in_state(
+        mut self,
+        _moder: &mut MODER<P>,
+        _otyper: &mut OTYPER<P>,
+        initial_state: PinState,
+    ) -> Pin<Output<OpenDrain>, HL, P, N> {
+        self._set_state(initial_state);
+        self.mode::<Output<OpenDrain>>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as an push pull output pin
+    /// Initial state will be low.
+    pub fn into_push_pull_output(
+        mut self,
+        _moder: &mut MODER<P>,
+        _otyper: &mut OTYPER<P>,
+    ) -> Pin<Output<PushPull>, HL, P, N> {
+        self._set_low();
+        self.mode::<Output<PushPull>>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as an push-pull output pin.
+    /// `initial_state` specifies whether the pin should be initially high or low.
+    pub fn into_push_pull_output_in_state(
+        mut self,
+        _moder: &mut MODER<P>,
+        _otyper: &mut OTYPER<P>,
+        initial_state: PinState,
+    ) -> Pin<Output<PushPull>, HL, P, N> {
+        self._set_state(initial_state);
+        self.mode::<Output<PushPull>>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as an analog input pin
+    pub fn into_analog(
+        mut self,
+        _moder: &mut MODER<P>,
+        _pupdr: &mut PUPDR<P>,
+    ) -> Pin<Analog, HL, P, N> {
+        self.mode::<Analog>();
+        Pin::new()
+    }
+
+    /// Configures the pin to operate as an touch sample
+    pub fn into_touch_sample(
+        self,
+        moder: &mut MODER<P>,
+        otyper: &mut OTYPER<P>,
+        afr: &mut Afr<HL, P>,
+    ) -> Pin<Alternate<OpenDrain, 9>, HL, P, N> {
+        self.into_alternate_open_drain(moder, otyper, afr)
+    }
+
+    /// Configures the pin to operate as an touch channel
+    pub fn into_touch_channel(
+        self,
+        moder: &mut MODER<P>,
+        otyper: &mut OTYPER<P>,
+        afr: &mut Afr<HL, P>,
+    ) -> Pin<Alternate<PushPull, 9>, HL, P, N> {
+        self.into_alternate(moder, otyper, afr)
+    }
```

Comment:
I would personally remove those because they are just convenience methods. They have been there before but such a change in the pin types seems like a nice chance to clean this up.